### ISSUE TITLE
introduce filterSystem

### ISSUE THREAD

**Issue by blaggacao** [numtide/flake-utils#27]
As an alternative to `flattenTreeSystem`, I propose `filterSystem` that directly acts on the output of current `flattenTree` and filters with a predicate that drops all packages that would _trivially_ fail a `nix flake check`.

/cc @pacman99 @nrdxp

**Comment by Pacman99** [comment]
My solution but with nixpkgs lib:
```nix
 filterPackages = system: packages:
    let
      # Everything that nix flake check requires for the packages output
      filter = (n: v: with v; let platforms = meta.hydraPlatforms or meta.platforms or [ ]; in
      lib.isDerivation v && !meta.broken && builtins.elem system platforms);
    in
    lib.filterAttrs filter packages;
 ```

**Comment by nrdxp** [comment]
If it wouldn't be too much work to adapt @Pacman99's example to a version without a dependancy on nixpkgs, I think filtering would be a sane way to allow users to start experimenting with the newer `packages` output, without having to manually declare each system indivdually. 

Obviously `flattenTreeSystem` got to a point where it was too complex too easily follow and debug, but perhaps just filtering would be more managable. If I ever find the time (which seems like a bigger and bigger if at this point), perhaps I could try my hand at a nixpkgsless implementaton.

**Comment by zimbatm** [comment]
When I wrote `flattenTree`, I was under the impression that we should use `packages` and avoid `legacyPackages` due to the naming. Since then, my thinking has changed; nixpkgs is unlikely to become flat, as @nrdxp observed. I also think that having a hierarchy is useful for monorepos.

So for now, I would suggest ignoring the naming and use `legacyPackages` instead. From what I know about flakes, I really hope that this is not the final design as there are some other deeper issues with it. For example, if each flake creates an instance of nixpkgs, and evaluating nixpkgs takes 400MB of RAM, it's not going to scale.

**Comment by blaggacao** [comment]
Implemented by #28
